File: java/algs4/chap1/1.1.20.java
public static float ln(long N) {
		float lo = 0;
		float hi = N;
		while (lo < hi) {
			float mid = (lo + hi)/2;
			if ((mid*mid) < N)
				lo = mid;
			if ((mid*mid) > N)
				hi = mid;
			if (Math.abs(mid*mid - N) < 0.1)
				return mid;
		}
		return -1;
	}
	
	public static long recursive(long N) {
		return recursiveReal(N,1);
	}
	
	public static long recursiveReal(long N, long result) {
		if (N == 1) {
			return result;
		}
		if (N > 1) {
			result *= N;
			return recursiveReal(N-1, result);
		}
		return result;//This return can't be removed.It's confussing.
	}

	public static void main(String[] args) {
		int N = StdIn.readInt();
		StdOut.println(ln(recursive(N)));

	}

File: python/algs4/chap1/1.1.20.py
def ln(N):
	lo = 0
	hi = N
	while lo < hi:
		mid = (lo+hi)/2
		if mid**2 < N:
			lo = mid
		if mid**2 > N:
			hi = mid
		if abs(mid**2 - N) < 0.01:
			return mid
	return 'Error'

def recursive(N):
    return recursive1(N,1)

def recursive1(N, result):
	if N == 1:
		return result
	if N > 1:
		result *= N
		return recursive1(N-1, result)

if __name__ == "__main__":
	N = (float)(input())
	print (ln(recursive(N)))
